protobuf/avs.proto

```diff
@@ -136,115 +131,173 @@ enum TaskStatus {
+message GraphQLQueryNode {
   // TODO: support graphql variable
   string url   = 1;
   string query = 2;
 }
 
-message HTTPAPICall {
+message RestAPINode {
   string              url     = 1;
   map<string, string> headers = 2;
   string              body    = 3;
+  string              method  = 4;
 }
 
 enum CustomCodeType {
   JavaScript = 0;
 }
 
-message CustomCode {
+message CustomCodeNode {
   CustomCodeType type     = 1;
-  string          body    = 2;
+  string         source    = 2;
 }
 
 message ConditionJump {
   string expression  = 1;
   string next        = 2;
 }
 
-message BranchAction {
-  ConditionJump If                  = 1;
-  repeated ConditionJump ElseIfs    = 2;
-  ConditionJump Else                = 3;
+message BranchNode {
+  ConditionJump if                  = 1;
+  repeated ConditionJump elseIfs    = 2;
+  ConditionJump else                = 3;
 }
 
-message TaskAction {
-  TaskType task_type = 1;
+message FilterNode {
+  // Filter node acts like .select or .filter to pluck out element in an array that evaluate the expression to true
+  string expression = 1;
+}
 
-  string id = 2;
-  string name = 3;
+// The edge is relationship or direct between node
+message TaskEdge {
+  string id     = 1 ;
+  string source = 2 ;
+  string target = 3 ; 
+}
+
+// TaskType represents what kind of work the task will perform
+enum TaskType {
+  // Handle default/missing data
+  ETHTransferTask = 0;
+  
+  // Generic contract execution which can be used for:
+  // ERC20 Transfer, NFT Transfer, auto reclaim, auto restaking etc
+  // When executing a contract we need at least 2 things:
+  // - target contract address
+  // - the message to send to that contract
+  ContractWriteTask = 1;
+  ContractReadTask  = 2;
+
+  // GraphQL query. Note that a graphql can also be query as a HTTP API, but having its own graphql type make writing query easiser without passing all the data in post as parameter.
+  GraphQLDataQueryTask = 3;
 
-  // Next can be empty. In some kind of block, such as branching, the next is
-  // based on branching condition
-  repeated string next = 4;
-
-  // Transfer eth
-  ETHTransfer eth_transfer = 10;
-  // Run one ore more contracts. The call call also be batched with tool like
-  // multicall to wrap many calls
-  ContractExecution contract_execution = 11;
-  // Make call to a graphql endpoint
-  GraphQLDataQuery graphql_data_query = 12;
   // Make call to a HTTP endpoint
-  HTTPAPICall http_data_query = 13;
+  RestAPITask = 4;
+
   // CustomCode allow to run arbitraty JavaScript.
-  CustomCode custom_code = 14;
-  BranchAction branch = 15;
+  BranchTask = 5;
+  FilterTask = 6;
+  CustomCodeTask = 7;
+}
+
+message TaskNode {
+  TaskType node_type = 1;
+  string id   = 2;
+  string name = 3;
+
+  // based on node_type one and only one of these field are set
+  oneof task_body {
+    // Transfer eth require no calldata etc, just a destination address and an eth amount to be sent
+    ETHTransferNode   eth_transfer = 10;
```

Comment:
Regarding the `node_type: 'ETHTransferTask'`, I’d just call it `Transfer` to represent the default transfer event on EVM chains, because the `ETH` part is already implied.

Reply:
Hmm, that’s true. Okay let’s keep the naming conprehensive.